common/expression/src/evaluator.rs

```diff
@@ -307,6 +127,277 @@ impl Evaluator {
+        &self,
+        span: Span,
+        column: Column,
+        dest_type: &DataType,
+    ) -> Result<Column> {
+        match (column, dest_type) {
+            (Column::Null { len }, DataType::Nullable(_)) => {
+                let mut builder = ColumnBuilder::with_capacity(dest_type, len);
+                for _ in 0..len {
+                    builder.push_default();
+                }
+                Ok(builder.build())
+            }
+            (Column::EmptyArray { len }, DataType::Array(_)) => {
+                let mut builder = ColumnBuilder::with_capacity(dest_type, len);
+                for _ in 0..len {
+                    builder.push_default();
+                }
+                Ok(builder.build())
+            }
+            (Column::Nullable { column, validity }, DataType::Nullable(dest_ty)) => {
+                let column = self.run_cast_column(span, *column, dest_ty)?;
+                Ok(Column::Nullable {
+                    column: Box::new(column),
+                    validity,
+                })
+            }
+            (col, DataType::Nullable(dest_ty)) => {
+                let column = self.run_cast_column(span, col, dest_ty)?;
+                Ok(Column::Nullable {
+                    validity: constant_bitmap(true, column.len()).into(),
+                    column: Box::new(column),
+                })
+            }
+            (Column::Array { array, offsets }, DataType::Array(dest_ty)) => {
+                let array = self.run_cast_column(span, *array, dest_ty)?;
+                Ok(Column::Array {
+                    array: Box::new(array),
+                    offsets,
+                })
+            }
+            (Column::Tuple { fields, len }, DataType::Tuple(fields_ty)) => {
+                let new_fields = fields
+                    .into_iter()
+                    .zip(fields_ty)
+                    .map(|(field, field_ty)| self.run_cast_column(span.clone(), field, field_ty))
+                    .collect::<Result<_>>()?;
+                Ok(Column::Tuple {
+                    fields: new_fields,
+                    len,
+                })
+            }
+
+            // identical types
+            (col @ Column::Null { .. }, DataType::Null)
+            | (col @ Column::EmptyArray { .. }, DataType::EmptyArray)
+            | (col @ Column::Boolean(_), DataType::Boolean)
+            | (col @ Column::String { .. }, DataType::String) => Ok(col),
+
+            (col, dest_ty) => {
+                // number types
+                with_number_type!(SRC_TYPE, match &col {
+                    Column::SRC_TYPE(col) => {
+                        with_number_type!(DEST_TYPE, match dest_ty {
+                            DataType::DEST_TYPE => {
+                                let mut new_col = Vec::with_capacity(col.len());
+                                for &val in col.iter() {
+                                    let new_val = val.try_into().map_err(|_| {
```

Comment:
`try_into` may cause too much branch prediction.

Can we have a faster path for `lower_bits` to `higher_bits` ?

There is `AsPrimitive` trait.

Reply:
Are you suggesting adding a wrapping mode for the cast?

There are three situations when a cast will present: 

1. When `type_check::can_auto_cast_to` agree to accept an argument with a not totally matched type.
2. When the user uses `CAST` in SQL.
3. When the user uses `TRY_CAST` in SQL.

And also, we could provide four different ways to handle an unsuccessful cast:

1. Wrap it.
2. Make it Null.
3. Give a warning.
4. Throw an error and stop the execution.

My question is what behavior do we use for each different cast situation? And also, @leiysky is suggesting abandoning the fourth behavior (throwing error) because it will be quite unexpected to the user when an error kills a query that has been running for a whole day.

Reply:
> Are you suggesting adding a wrapping mode for the cast?

Not yet, I was suggesting to improve the performance when casting from low-range number to high-range number.

https://togithub.com/jorgecarleitao/arrow2/issues/252

> And also, we could provide four different ways to handle an unsuccessful cast:

For number cast to number, if overflow happens, we can result in the first behavior (as clickhouse) or the fourth behavior(as duckdb,PostgreSQL) .

If users do not want errors, they can use `try_cast`.

Reply:
addressed